### PR TITLE
[OpenVINO] Fix nn ops correctness: moments, binary_crossentropy, softmax, log_softmax

### PR DESCRIPTION
**Pull request by goyaladitya05**
This PR fixes multiple correctness issues in the OpenVINO `nn.py` backend and re-enables previously excluded tests.

## Fixes
- **`softmax` / `log_softmax`**: Added support for tuple/list axis by iterating reductions across sorted axes using a numerically stable implementation.
- **`binary_crossentropy` (`from_logits`)**: Replaced sigmoid-based computation with a stable softplus formulation to avoid `inf` / `nan`.
- **`moments`**: Prevented float16 overflow by computing reductions in float32 and casting results back.

## Tests Re-enabled
- NNOpsCorrectnessTest::test_softmax_correctness_with_axis_tuple
- NNOpsCorrectnessTest::test_log_softmax_correctness_with_axis_tuple
- NNOpsCorrectnessTest::test_moments
- NNOpsCorrectnessTest::test_normalize
- NNOpsCorrectnessTest::test_rms_normalization_10.0
- BinaryCrossentropyTest::test_label_smoothing

Closes: openvinotoolkit/openvino/issues/34417

### CODE REVIEW

File: keras/src/backend/openvino/nn.py [hunk 1]
```diff
@@ -698,20 +751,31 @@ def binary_crossentropy(target, output, from_logits=False):
 
 
 def moments(x, axes, keepdims=False, synchronized=False):
+    import numpy as np
+
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/openvino/nn.py]: removed.

File: keras/src/backend/openvino/nn.py [hunk 2]
```diff
@@ -169,20 +172,59 @@ def softmax(x, axis=-1):
...
+            True,
+        ).output(0)
+    x_shifted = ov_opset.subtract(x, x_max).output(0)
+    sum_exp = ov_opset.exp(x_shifted).output(0)
+    for ax in axes:
+        sum_exp = ov_opset.reduce_sum(
+            sum_exp,
+            ov_opset.constant([ax], Type.i32).output(0),
+            True,
+        ).output(0)
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/openvino/nn.py]: done

File: keras/src/backend/openvino/nn.py [hunk 3]
```diff
@@ -169,20 +172,59 @@ def softmax(x, axis=-1):
...
+        axes = sorted(axis)
+    else:
+        axes = [axis]
+    x_max = x
+    for ax in axes:
+        x_max = ov_opset.reduce_max(
+            x_max,
+            ov_opset.constant([ax], Type.i32).output(0),
+            True,
+        ).output(0)
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/openvino/nn.py]: done

File: keras/src/backend/openvino/nn.py [hunk 4]
```diff
@@ -169,20 +172,59 @@ def softmax(x, axis=-1):
...
+                True,
+            ).output(0)
+        exp_x = ov_opset.exp(ov_opset.subtract(x, x_max).output(0)).output(0)
+        sum_exp = exp_x
+        for ax in axes:
+            sum_exp = ov_opset.reduce_sum(
+                sum_exp,
+                ov_opset.constant([ax], Type.i32).output(0),
+                True,
+            ).output(0)
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/openvino/nn.py]: done

File: keras/src/backend/openvino/nn.py [hunk 5]
```diff
@@ -169,20 +172,59 @@ def softmax(x, axis=-1):
...
         )
+    if isinstance(axis, (tuple, list)):
+        axes = sorted(axis)
+        x_max = x
+        for ax in axes:
+            x_max = ov_opset.reduce_max(
+                x_max,
+                ov_opset.constant([ax], Type.i32).output(0),
+                True,
+            ).output(0)
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/openvino/nn.py]: done

File: keras/src/backend/openvino/nn.py [hunk 6]
```diff
@@ -169,20 +170,38 @@ def softmax(x, axis=-1):
...
+    x_max = ov_opset.reduce_max(x, axes_const, True).output(0)
+    x_shifted = ov_opset.subtract(x, x_max).output(0)
+    sum_exp = ov_opset.reduce_sum(
+        ov_opset.exp(x_shifted).output(0), axes_const, True
+    ).output(0)
+    log_sum_exp = ov_opset.log(sum_exp).output(0)
+    result = ov_opset.subtract(x_shifted, log_sum_exp).output(0)
+    if restore_shape is not None:
+        result = ov_opset.reshape(result, restore_shape, False).output(0)
+    return OpenVINOKerasTensor(result)
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/openvino/nn.py:208]: Added

File: keras/src/backend/openvino/nn.py [hunk 7]
```diff
@@ -169,20 +170,38 @@ def softmax(x, axis=-1):
         return OpenVINOKerasTensor(
             ov_opset.reshape(softmax_x, x_shape, False).output(0)
         )
+    if isinstance(axis, (tuple, list)):
+        axes_const = ov_opset.constant(sorted(axis), Type.i32).output(0)
+        x_max = ov_opset.reduce_max(x, axes_const, True).output(0)
+        exp_x = ov_opset.exp(ov_opset.subtract(x, x_max).output(0)).output(0)
+        sum_exp = ov_opset.reduce_sum(exp_x, axes_const, True).output(0)
+        return OpenVINOKerasTensor(ov_opset.divide(exp_x, sum_exp).output(0))
```
[the remark this answers is not in the document]
Comment (goyaladitya05) [reply keras/src/backend/openvino/nn.py:180]: Added

**Comment by goyaladitya05** [comment]
I've addressed all the gemini reviews, and the CI tests are passing. Ready for review.

File: keras/src/backend/openvino/numpy.py [hunk 8]
```diff
@@ -573,6 +574,10 @@ def argmin(x, axis=None, keepdims=False):
 
 
 def argsort(x, axis=-1):
+    # Non-OV inputs (e.g. plain lists from ops/nn.py tuple-axis dispatch)
+    # must not enter the OV graph path to avoid constant-folding deadlocks.
+    if not isinstance(x, (OpenVINOKerasTensor, ov.Output, KerasVariable)):
+        return np.argsort(x, axis=axis).astype(np.int32)
```
Comment (hertschuh) [inline keras/src/backend/openvino/numpy.py]: This should be reverted.

The bug is that `nn.py` should not use `backend.numpy.argsort`.

Is that something you can fix? You can use this trick to do the argsort: https://stackoverflow.com/a/3382369

It's a short list so there is no benefit to using numpy, let's use plain Python.

Comment (goyaladitya05) [reply keras/src/backend/openvino/numpy.py]: Done. I've implemented the trick directly to `keras/src/ops/nn.py`.
